oak_client/src/attestation.rs

```diff
@@ -0,0 +1,156 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use anyhow::{anyhow, Context};
+use http::uri::Uri;
+use tokio_rustls::rustls;
+use tonic::transport::{Channel, ClientTlsConfig};
+use x509_parser::{
+    der_parser::{self, der::parse_der, oid::Oid},
+    parse_x509_certificate,
+};
+
+// Using [`Nid::NETSCAPE_COMMENT`] identifier since `rust-openssl` doesn't support custom
+// extensions yet.
+// https://github.com/sfackler/rust-openssl/issues/1411
+// https://www.alvestrand.no/objectid/2.16.840.1.113730.1.13.html
+const TEE_EXTENSION_OID: Oid<'static> = der_parser::oid!(2.16.840.1.113730.1.13);
+
+/// Creates a remotely attested TLS channel for connecting to Oak.
+pub async fn create_attested_tls_channel(
+    uri: &Uri,
+    root_tls_certificate: &[u8],
+    tee_quote: &str,
+) -> anyhow::Result<Channel> {
+    let tls_config = create_tls_attestation_config(root_tls_certificate, tee_quote)
+        .context("Couldn't create TLS config")?;
+    Channel::builder(uri.clone())
+        .tls_config(tls_config)
+        .context("Couldn't create TLS configuration")?
+        .connect()
+        .await
+        .context("Couldn't connect to Oak Application")
+}
+
+/// Create client TLS configuration with a custom X.509 certificate verifier.
+/// [`root_tls_certificate`] should contain only a single PEM encoded certificate.
+/// https://tools.ietf.org/html/rfc1421
+pub fn create_tls_attestation_config(
+    root_tls_certificate: &[u8],
+    tee_quote: &str,
+) -> anyhow::Result<ClientTlsConfig> {
+    let mut config = rustls::ClientConfig::new();
+
+    // Enable HTTP/2.
+    const ALPN_H2: &str = "h2";
+    config.set_protocols(&[Vec::from(&ALPN_H2[..])]);
+
+    // Add root TLS certificate.
+    let mut cc_reader = std::io::BufReader::new(&root_tls_certificate[..]);
+    let certs = rustls::internal::pemfile::certs(&mut cc_reader)
+        .map_err(|error| anyhow!("Couldn't parse TLS certificate: {:?}", error))?;
+    if certs.len() != 1 {
+        return Err(anyhow!("Only a single PEM certificate should be provided"));
+    }
+    config
+        .root_store
+        .add(&certs[0])
+        .context("Couldn't add root certificate")?;
+
+    // Set custom X.509 certificate verifier.
+    config
+        .dangerous()
+        .set_certificate_verifier(std::sync::Arc::new(RemoteAttestationVerifier::new(
+            tee_quote,
+        )));
+
+    Ok(ClientTlsConfig::new().rustls_client_config(config))
+}
+
+/// Custom X.509 certificate verifier that checks TEE quotes in the X.509 TEE extension.
+struct RemoteAttestationVerifier {
+    tee_quote: String,
```

Comment:
I think in general we would not know the exact quote to expect, since it contains extra data and signatures and stuff. So passing the literal string here seems incorrect. I think what we need to pass is some sort of predicate / verifier for the quote itself, which for now may just be a `Fn(&str) -> QuoteVerified` or something like that. Then for now we could have a single implementation that just checks for equivalence, it can be an anonymous lambda for that.

edit: I guess this is actually what `RemoteAttestationVerifier` actually is, but I think it should be defined as a trait, allow using a lambda for the check logic, clarify what the return value means, and clarify what the parameter is (`extension` seems a bit vague; also isn't that extension always the quote? or can there be other stuff in it? also `TEE_EXTENSION_OID` suggests that this is the `tee_extension` value, so maybe just stick to this terminology, and avoid quote? or use `tee_quote_*` everywhere)

Reply:
I think having it as a trait would require `#![feature(associated_type_bounds)]`, since we would need to implement `trait rustls::ServerCertVerifier` for this trait.

https://github.com/rust-lang/rfcs/pull/2289/files#diff-6cd727664c55412d02e308dfb2dcf5f5ee273a0c6f728160f79e68e1fab88e7bR71-R78

Reply:
Currently made a `Quote` a separate decodable class
And this verifier takes `tee_measurement` as an argument

Reply:
What I mean is that I think this class should only hold something like `Predicate<Report>` or similar (e.g. by using https://crates.io/crates/predicates , or anyways a fn from Report to bool). Then the caller can decide how to build one of those. I think this would make more sense if you expand the Report struct to be more realistic, for instance by adding some more fields, similar to https://pkg.go.dev/github.com/edgelesssys/ego@v0.1.0/attestation#Report . Then there may be different validation logic for each field or combination of fields.

Reply:
Created an issue: https://github.com/project-oak/oak/issues/1896